x/twap/logic_test.go

```diff
@@ -42,76 +43,287 @@ func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
-		updateTime time.Time
-		expRecord  types.TwapRecord
+		recordsToPreSet     types.TwapRecord
+		testPoolId          uint64
+		testDenom0          string
+		testDenom1          string
+		testTime            time.Time
+		expectedAccumulator sdk.Dec
+		expectedErr         bool
 	}{
-		"0 accum start": {
-			record:     newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
-			updateTime: time.Unix(2, 0),
-			expRecord:  newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		"same time with existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime,
+		},
+		"call 1 second after existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime.Add(time.Second),
+			// 1000000000(spot price) * 1(time)
+			expectedAccumulator: baseRecord.P0ArithmeticTwapAccumulator.Add(sdk.NewDec(1000000000)),
+		},
+		"call 1 second before existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime.Add(-time.Second),
+			expectedErr:     true,
+		},
+		"on lexicographical order denom parameters": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset1Denom,
+			testDenom1:      baseRecord.Asset0Denom,
+			testTime:        baseTime,
+		},
+		"test non lexicographical order parameter": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId + 1,
+			testDenom0:      baseRecord.Asset1Denom,
+			testDenom1:      baseRecord.Asset0Denom,
+			testTime:        baseTime,
+			expectedErr:     true,
 		},
 	}
+
 	for name, test := range tests {
 		s.Run(name, func() {
-			// setup common, block time, pool Id, expected spot prices
-			s.Ctx = s.Ctx.WithBlockTime(test.updateTime.UTC())
-			test.record.PoolId = poolId
-			test.expRecord.PoolId = poolId
-			test.expRecord.P0LastSpotPrice = newSp
-			test.expRecord.P1LastSpotPrice = newSp
-			test.expRecord.Height = s.Ctx.BlockHeight()
-			test.expRecord.Time = s.Ctx.BlockTime()
-
-			newRecord := s.twapkeeper.UpdateRecord(s.Ctx, test.record)
-			s.Require().Equal(test.expRecord, newRecord)
+			s.SetupTest()
+			s.twapkeeper.StoreNewRecord(s.Ctx, test.recordsToPreSet)
+
+			interpolatedRecord, err := s.twapkeeper.GetInterpolatedRecord(s.Ctx, test.testPoolId, test.testDenom0, test.testDenom1, test.testTime)
+			if test.expectedErr {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+
+			if test.testTime.Equal(baseTime) {
+				s.Require().Equal(test.recordsToPreSet, interpolatedRecord)
+			} else {
+				s.Require().Equal(test.testTime, interpolatedRecord.Time)
```

Comment:
Just wondering why not assert the entire twapRecord instead of individual values?

Reply:
Good question, we don't want to test the whole field of twapRecord here(some aren't initialized in accordance to the testing env, nor do we need to initialize them to test them here). Thats why we test and check only the fields that needs testing here!

Reply:
aah gotchu